SimpleUI: add oauth2 support

zeapo opened the pull request:
With these changes one can now authenticated on the simple-ui using oauth2.

The rights checks have been mutualized with the basic auth mechanism. 

For example, using azure, one simply has to add the following env environment:
```
HAWKBIT_SERVER_SECURITY_OAUTH2_CLIENT_ENABLED=true

SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_CLIENT_NAME=Login With Azure
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_PROVIDER=azure
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_CLIENT_ID=....
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_CLIENT_SECRET=...
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_AUTHORIZATION_GRANT_TYPE=...
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_REDIRECT_URI=...
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_AZURE_SCOPE=openid,profile,email
SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_AZURE_ISSUER_URI=....
SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_AZURE_JWK_SET_URI=....
```

# NOTE
~~An important drawback of the way the interceptor is made is that when the token expires there is no simple way to refresh it. As this is used too early and the user is not fully authenticated. Trying to rely on OAuth2AuthorizedClientManager leads nowhere as it wont find the client, hence will be unable to authenticate the request.~~

~~I'll probably open a future PR to update the way we obtain the authorities; this PR at least allows the user to work for 1h :)~~

Edit: it's fixed and is able to refresh the id token now

zeapo commented:
Switching to draft as I found a working solution that allows a refresh

avgustinmm commented on hawkbit-simple-ui/src/main/java/org/eclipse/hawkbit/ui/simple/security/SecurityConfiguration.java:
```diff
@@ -11,17 +11,26 @@
 @EnableWebSecurity
 @Configuration
 public class SecurityConfiguration extends VaadinWebSecurity {
+    private Customizer<OAuth2LoginConfigurer<HttpSecurity>> oAuth2LoginConfigurerCustomizer;
```
Please use a empty line after the class declaration and the fields/methods as in other places of in hawkBit. Have you applied the hawkbit code format?

zeapo replied:
I just noticed that it was not applied. sorry.

strailov commented on hawkbit-simple-ui/src/main/java/org/eclipse/hawkbit/ui/simple/security/OAuth2TokenManager.java:
```diff
@@ -0,0 +1,72 @@
+     * Tries to refresh the id token if it is expired and adds it to the request.
+     */
+    private String authorizedToken(OAuth2AuthenticationToken authentication) {
+        String registrationId = authentication.getAuthorizedClientRegistrationId();
```
Make variables final whenever possible.

zeapo commented:
I added the same mechanims that was added to mgmt server;  the boolean `hawkbit.server.security.oauth2.client.enabled` makes it simpler to enable / disable rather than checking for the client registery

added some null checks; tested with above boolean enabled / disabled; with registration configured / non configured. 
Behavior is:
- if enabled and no config > spring will complain that the registration is missing
- if disabled > with and without config is the same, will use basic auth

zeapo commented:
Please let me know if any further changes are required. Knowing that once Spring Security 6.5 is out and we bump version, I'll revisit this to remove the manual token refresh.

strailov commented:
I'll try to check this out one more time today and if everything is fine I think we are ready to merge it.

strailov commented:
Looks good, thanks @zeapo !
